src/gui/widget/TextField.cpp

```diff
@@ -0,0 +1,210 @@
+			maxTextWidth()
+		) {
+			return true;
+		}
+
+		sf::String tmp(text.getString());
+		tmp.insert(cursorIndex, e.text.unicode);
+		text.setString(tmp);
+		++cursorIndex;
+
+		return true;
+	}
+
+	if (e.type == sf::Event::KeyPressed) {
+		switch (e.key.code) {
+			case sf::Keyboard::BackSpace: {
+				if (cursorIndex == 0) {
+					return true;
+				}
+
+				--cursorIndex;
+
+				sf::String tmp(text.getString());
+				tmp.erase(cursorIndex);
+				text.setString(tmp);
+
+				return true;
+			}
+			case sf::Keyboard::Left:
+				if (cursorIndex != 0) {
+					--cursorIndex;
+				}
+				return true;
+			case sf::Keyboard::Right:
+				if (cursorIndex != text.getString().getSize()) {
+					++cursorIndex;
+				}
+				return true;
+			default:
+				return false;
+		}
+	}
+
+	if (e.type == sf::Event::MouseButtonPressed) {
+		if (e.mouseButton.button != sf::Mouse::Left) {
+			return false;
+		}
+
+		sf::Vector2f mousePosition = window.mapPixelToCoords(
+			sf::Vector2i(e.mouseButton.x, e.mouseButton.y)
+		);
+
+		if (!position.contains(mousePosition)) {
+			return false;
+		}
+
+		cursorIndex = getIndex(mousePosition.x - position.left);
+
+		return true;
+	}
+
+	return false;
+}
+
+unsigned int GUI::Widget::TextField::getIndex(const float& xPosition) const {
+	if (xPosition < textStartPos().x) {
+		return 0;
+	} else if (xPosition > text.getLocalBounds().width + textStartPos().x) {
+		return text.getString().getSize();
+	}
+
+        int a = 0;
+        int b = text.getString().getSize() - 1;
+	while (a <= b) {
+		unsigned int current = (a + b) / 2;
+
+		float charWidth = textWidth({text.getString()[current]});
+		float charPos = text.findCharacterPos(current).x;
+
+		if (std::abs(charWidth / 2 + charPos - xPosition) < charWidth) {
+			if (charWidth / 2 + charPos - xPosition < 0) {
+				return current + 2;
+			}
+			return current + 1;
+		}
+
+		if (charPos > xPosition) {
+			b = current - 1;
+		}
+		if (charPos < xPosition) {
+			a = current + 1;
+		}
+	}
+
+	return a + 1;
+}
+
+unsigned int GUI::Widget::TextField::fontSize() const {
+	return position.height * 0.8f - borderSize() * 2;
+}
+
+unsigned int GUI::Widget::TextField::borderSize() const {
+	return (position.height < 60 ? (position.height < 40 ? 2 : 4) : 6);
+}
+
+sf::Vector2f GUI::Widget::TextField::textStartPos() const {
+	return sf::Vector2f(
+		borderSize() + characterSize().x / 2,
+		position.height / 2 - characterSize().y / 2 - borderSize() / 2
+	);
+}
+
+float GUI::Widget::TextField::maxTextWidth() const {
+	return (
+		position.width -
+		borderSize() * 2 -
+		textStartPos().x -
+		characterSize().x / 2
+	);
+}
+
+float GUI::Widget::TextField::textWidth(const sf::String& str) const {
+	return sf::Text(str, font, fontSize()).getLocalBounds().width;
+}
+
+sf::Vector2f GUI::Widget::TextField::characterSize() const {
+	sf::Text tmp("j", font, fontSize());
+
+	sf::Vector2f size;
+	size.x = tmp.getLocalBounds().width;
+
+	tmp.setString(stringFromUtf8(u8"|_ÖÄ%£*"));
```

Comment:
Any specific reason to use characters other than those in Label? Would it be possible to re-use the Label code?